### PR TITLE
Update for bevy master

### LINKED ISSUE #23: Build fails with recent bevy master
due to https://github.com/bevyengine/bevy/pull/1236

```
error[E0046]: not all trait items implemented, missing: `copy_texture_to_buffer`, `copy_texture_to_texture`
  --> /Users/robparrett/.cargo/registry/src/github.com-1ecc6299db9ec823/bevy_webgl2-0.4.2/src/renderer/webgl2_render_context.rs:29:1
   |
29 | impl RenderContext for WebGL2RenderContext {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `copy_texture_to_buffer`, `copy_texture_to_texture` in implementation
   |
   = help: implement the missing item: `fn copy_texture_to_buffer(&mut self, _: TextureId, _: [u32; 3], _: u32, _: BufferId, _: u64, _: u32, _: Extent3d) { todo!() }`
   = help: implement the missing item: `fn copy_texture_to_texture(&mut self, _: TextureId, _: [u32; 3], _: u32, _: TextureId, _: [u32; 3], _: u32, _: Extent3d) { todo!() }`

error: aborting due to previous error
```

### PR DESCRIPTION
**Pull request by rparrett**
Fixes #23.

But in a pretty half-assed way. I do not know how to implement the render/render resource context stuff, so they are just stubs.

Providing this for anyone else who's tracking bevy master and just wants to get this working again.

CI obviously failing because this won't work with bevy 0.4.

### CODE REVIEW

**Comment by rparrett** [comment]
Up to date with current bevy master, but the new wgpu stuff was rough on me and I've done the bare minimum only to get my 2d game working.

**Comment by rparrett** [comment]
Oops! Thanks for that.

I'm super happy to keep working on this. Some of the examples are looking to be in pretty bad shape now that I've actually even run them. 

The biggest known problem I have is here: https://github.com/mrk-its/bevy_webgl2/pull/24/commits/d30ac8ee23a78dc987425e6e7a1c9409e9b3586a#diff-a731f7f5f708022a34572ab832954c802f7fd77de1e8c38418ed7049717912f7R150

Which is resulting in a glGenerateMipmap / GL_INVALID_OPERATION in my project (but no visible problems weirdly), but surely is causing all sorts of other issues.

**Comment by rparrett** [comment]
Here's a survey of the examples:

font_atlas_debug - only partially working
wasm.js:338 panicked at 'called `Option::unwrap()` on a `None` value', src/webgl2_render_pass.rs:227:89

3d_scene - camera is wonky. otherwise ok
load_gltf - nothing appears, GL_INVALID_OPERATION : glGenerateMipmap: 
parenting - camera is wonky. otherwise ok
texture - camera is wonky. otherwise ok?
z_sort_debug - cubes appear completely black

mesh_custom_attribute - camera is wonky. otherwise ok.
shader_custom_material - camera is wonky. otherwise ok
shader_defs - can't tell if ok because camera is too wonky

contributors - current contributor's sprite is not always highlighted?

**Comment by rparrett** [comment]
Oops. I wanted PerspectiveCameraBundle, not a 3d orthographic camera. Will fix that up.

**Comment by rparrett** [comment]
I am not sure why I've been manually fixing examples instead of just copying them and using `fix_bevy_example.sh`.

**Comment by rparrett** [comment]
Issue with the contributors example was fixed in bevy

**Comment by mrk-its** [comment]
Thanks, you've done a lot of work! I'll take a closer look tomorrow.

**Comment by mrk-its** [comment]
> **load_gltf**
> 
> ```
> appears to work, but GL_INVALID_OPERATION : glGenerateMipmap: in console
> ```

Do you still have this warning? I cannot replicate it

**Comment by rparrett** [comment]
I am (chrome 88, macos).

image

But my assumption that this is caused by my having basically removed support for anything but the default textureformat is a totally blind one.

I'm pretty sure that I don't remember seeing this with bevy 0.4 / bevy_webgl2 but I could be wrong.

**Comment by mrk-its** [comment]
The only place mipmaps are generated is https://github.com/mrk-its/bevy_webgl2/blob/637ca07c75ffbd29ab4661aed54f148eb11a39b4/src/renderer/webgl2_render_context.rs#L117 - It seems we can simply remove it because mipmaps are not used at all (because of hardcoded TEXTURE_MIN_FILTER = Gl::NEAREST). The whole texture filtering handling needs to be implemented yet - so let's remove this mipmap generation now (It would be great to run all texture-related examples after that, but I think everything should work exactly the same way)

**Comment by rparrett** [comment]
Cool. That seems to resolve the scary error message.

Any leads on what might be going on with `z_sort_debug`?

**Comment by mrk-its** [comment]
> Cool. That seems to resolve the scary error message. I double checked `load_gltf`, `sprite`, `sprite_sheet`, `texture` and they continue to work.
> 
> Any leads on what might be going on with `z_sort_debug`?

Not yet, will take a look

**Comment by mrk-its** [comment]
It seems z_order_debug in bevy's master is also broken, I've filled an issue: https://github.com/bevyengine/bevy/issues/1411

File: src/webgl2_render_pass.rs [hunk 1]
```diff
@@ -145,11 +147,8 @@ impl<'a> RenderPass for WebGL2RenderPass<'a> {
     }
 
     fn draw_indexed(&mut self, indices: Range<u32>, _base_vertex: i32, instances: Range<u32>) {
-        let (index_type, type_size) = match self.pipeline_descriptor.as_ref().unwrap().index_format
```
Comment (mrk-its) [inline src/webgl2_render_pass.rs:148]: let's remove WebGL2RenderPass.pipeline_descriptor, as it was used only for providing access to index_format.
We can store gl index_format / index_type_size in WebGL2Pipeline object (in set_index_buffer method) or directly in WebGL2RenderPass object

Comment (rparrett) [reply src/webgl2_render_pass.rs:148]: I had attempted moving the `IndexFormat` to `WebGL2Pipeline` at some point, but the code ended up in a state where I had duplicated this chunk of code from `setup_vao`

```
        let resources = &self.render_context.render_resource_context.resources;
        let mut pipelines = resources.pipelines.write();
        let pipeline_handle = self.pipeline.as_ref().unwrap();
        let pipeline = pipelines.get_mut(&pipeline_handle).unwrap();
```

And that bothered me. And even then I never got it to a point where things even showed up on the screen again so I gave up.

I'll give it another go though.

Comment (rparrett) [reply src/webgl2_render_pass.rs:148]: Ah yes, this is the deeply mysterious panic I was getting when I attempted this last time:

```
wasm.js:752 panicked at 'Parking not supported on this platform', /Users/robparrett/.cargo/registry/src/github.com-1ecc6299db9ec823/parking_lot_core-0.8.2/src/thread_parker/wasm.rs:26:9
```

Comment (rparrett) [reply src/webgl2_render_pass.rs:148]: Okay, I think I worked around that. Commit is incoming. But maybe you can suggest a cleaner way to do it.

Comment (mrk-its) [reply src/webgl2_render_pass.rs:148]: I think it is fine now, it may be refactored later if we find a better way.

**Comment by JeanMertz** [comment]
Nice work @rparrett. It looks like this PR is ready to review/merge, correct? cc @mrk-its.

**Comment by rparrett** [comment]
It's probably not a goal to get this merged until we're much closer to bevy 0.5 being released.

**Comment by JeanMertz** [comment]
Got it. Makes sense. I'll just point to this branch meanwhile 👍

**Comment by mrk-its** [comment]
@rparrett @JeanMertz I think it is good idea to merge it (if everything still works with current bevy master), and we can keep syncing with bevy in separate PR's. Some time ago I've created separate branch https://github.com/mrk-its/bevy_webgl2/tree/v0.4 for 0.4.x version of bevy_webgl2, so we can sync with bevy's master on master branch here.

**Comment by rparrett** [comment]
Looks like all `bevy_webgl2` examples that are expected to work (everything but `font_atlas_debug`) on my end.

**Pull request merged**